libosmscout/src/osmscout/async/ReadWriteLock.cpp

```diff
@@ -0,0 +1,156 @@
+  License as published by the Free Software Foundation; either
+  version 2.1 of the License, or (at your option) any later version.
+
+  This library is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
+ */
+
+#include <osmscout/private/Config.h>
+
+#include <osmscout/async/ReadWriteLock.h>
+
+#include <osmscout/log/Logger.h>
+
+#include <cassert>
+
+namespace osmscout {
+
+/**
+ * X flag is defined as the following:
+ * 0 : X is released
+ * 1 : X is held, but waits for release of S
+ * 2 : X is held
+ * 3 : X was released and left available for one of request in wait
+ */
+
+void Latch::lock() {
+  std::thread::id tid = std::this_thread::get_id();
+
+  spin_lock();
+
+  if (x_owner != tid) {
+    /* increments the count of request in wait */
+    ++x_wait;
+    for (;;) {
+      /* if flag is 0 or 3 then it hold X with no wait,
+       * in other case it have to wait for X gate
+       */
+      if (x_flag == 0) {
+        ++x_flag;
+        --x_wait;
+        break;
+      } else if (x_flag == 3) {
+        x_flag = 1;
+        --x_wait;
+        break;
+      } else {
+        /* !!! pop gate first then unlock spin (reverse order for S request) */
+        std::unique_lock<std::mutex> lk(x_gate_lock);
+        spin_unlock();
+        x_gate.wait(lk);
+      }
+      spin_lock();
+    }
+
+    /* X = 1, check the releasing of S */
+    for (;;) {
+      /* if the count of S is zeroed then it finalize with no wait,
+       * in other case it have to wait for S gate */
+      if (s_count == 0) {
+        ++x_flag;
+        break;
+      } else {
+        /* !!! pop gate first then unlock spin (reverse order for notifier),
+         * the lock can be held here by itself, or by the thread is releasing
+         * the last S */
+        std::unique_lock<std::mutex> lk(s_gate_lock);
+        spin_unlock();
+        s_gate.wait(lk);
+      }
+      spin_lock();
+    }
+
+    /* X = 2, set owner */
+    x_owner = tid;
+  }
+
+  spin_unlock();
+}
+
+void Latch::unlock() {
+  spin_lock();
+  if (x_owner == std::this_thread::get_id()) {
+    x_owner = std::thread::id();
+    /* if X priority is enabled then hand-over to a request in wait for X */
+    x_flag = (px && x_wait > 0 ? 3 : 0);
+    spin_unlock();
+    x_gate.notify_all();
+  } else {
+    spin_unlock();
+  }
+}
+
+void Latch::lock_shared() {
+  spin_lock();
+  if (x_owner != std::this_thread::get_id()) {
+    /* if flag is 0 or 1 then it hold S with no wait,
+     * in other case it have to wait for X gate
+     */
+    for (;;) {
+      if (x_flag.load() < 2) {
```

Comment:
when x_flag is one (`X is held, but waits for release of S`) should not be new shared lock waiting?

Reply:
Not at all ! if you do that it is a deadlock

Reply:
can you please elaborate it more? how can I get deadlock?

During my testing, I get to the state, when one thread is waiting for exclusive lock (`x_flag=1`), then shared lock was released and immediately acquired another shared lock... I would expect that second shared lock will wait until exclusive one is properly acquired and released...

Reply:
you are right, I get deadlock, but I am just blind - don't understand how :-/

Reply:
I want to solve this case:
```
    osmscout::Latch latch;
    std::atomic_int i=0;
    osmscout::ReadLock rl(latch);

    std::thread t([&latch, &i](){
      osmscout::WriteLock wl(latch);
      i++;
    });

    std::this_thread::sleep_for(1s);

    rl.unlock();
    rl.lock();

    assert(i==1);

    t.join();
```
this code may hit the assert... but it should be guaranteed that second shared lock will to be locked before exclusive one...

Reply:
Ah, I see... The deadlock will happen with re-entrant shared locking. Damn!

Reply:
for more explanation: blocking a shared lock which is by nature re-entrant you will go straight into a dead end == deadlock.